fontforge: update to 20200314

markmentovai opened the pull request:
The previous port, 20190801_2, was broken when built with Python:

```
% fontforge
dyld: Symbol not found: _PyBool_Type
  Referenced from: /opt/local/lib/libfontforge.3.dylib
  Expected in: flat namespace
 in /opt/local/lib/libfontforge.3.dylib
zsh: abort      fontforge
```

Since 20190801, FontForge abandoned its autotools-based build in favor
of CMake. The new CMake-based build does not share this problem.

I found a stale pull request, #6289, that was similar, but used an interim git hash to update to the CMake-based build in advance of there being a proper FontForge release. There has since been a release, so I’m submitting this pull request to both update to the latest release and to adopt the CMake-based build. This is also a more direct translation of the existing Portfile, allowing the CMake-based build’s defaults to stand where appropriate without explicit configuration.

#### Description

<!-- Note: it is best to make pull requests from a branch rather than from master -->

###### Type(s)
<!-- update (title contains ": U(u)pdate to"), submission (new Portfile) and CVE Identifiers are auto-detected, replace [ ] with [x] to select -->

- [ ] bugfix
- [ ] enhancement
- [ ] security fix

###### Tested on
<!-- Generate version information with this command in shell:
    echo "macOS $(sw_vers -productVersion) $(sw_vers -buildVersion)"; echo "Xcode $(xcodebuild -version | awk -v ORS=' ' '{print $NF}')"
-->
macOS 10.15.4 19E287
Xcode 11.4.1 11E503a 

###### Verification <!-- (delete not applicable items) -->
Have you

- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [x] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [ ] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL?
<!-- Please don't open a new Trac ticket if you are submitting a pull request. -->
- [x] checked your Portfile with `port lint`?
- [ ] tried existing tests with `sudo port test`?
- [x] tried a full install with `sudo port -vst install`?
- [x] tested basic functionality of all binary files?

<!-- Use "skip notification" (surrounded with []) to avoid notifying maintainers -->

mojca commented on graphics/fontforge/Portfile:
```diff
@@ -3,10 +3,12 @@
-github.setup        fontforge fontforge 20190801
-revision            2
+github.setup        fontforge fontforge 20200314
+revision            1
```
```suggestion
revision            0
```

mojca commented on graphics/fontforge/Portfile:
```diff
@@ -65,39 +67,34 @@ platform darwin 8 {
-    configure.args-append   am_cv_python_pythondir=${frameworks_dir}/Python.framework/Versions/3.7/lib/python3.7/site-packages \
-                            am_cv_python_pyexecdir=${frameworks_dir}/Python.framework/Versions/3.7/lib/python3.7/site-packages
+    configure.args-delete   -DENABLE_PYTHON_SCRIPTING=False \
+                            -DENABLE_PYTHON_EXTENSION=False
```
You should probably use `configure.args-replace` and set the value to `True`, else it may not always be reproducible.

markmentovai replied:
I didn’t do this initially because FontForge’s own build defaults to these being on, but at your suggestion, I made this explicit, and made the other options that the CMake build makes configurable but that MacPorts doesn’t (and doesn’t need to) provide knobs for explicit too.

mojca commented on graphics/fontforge/Portfile:
```diff
@@ -65,39 +67,34 @@ platform darwin 8 {
-    depends_lib-append      port:xorg-libXi port:gtk3
-    configure.args-replace  --without-x --with-x
-    configure.args-append   --enable-gdk=yes
+    depends_lib-append      port:gtk3
```
Does this port generate `FontForge.app` inside Applications?
It would be nice if it did, see how it was done in #6289.

markmentovai replied:
The CMake build has a `macbundle` target that makes `FontForge.app` ([here](https://github.com/fontforge/fontforge/blob/0686c0609632cbe876e8038871d70542108a8613/osx/CMakeLists.txt)). It appears somewhat more complete than the one in #6289. For example, it has a more complete `Info.plist` that has file associations and icons. Problem is, that target doesn’t work at all (at least not without `realpath`, so `coreutils` would be a build-time dependency).

I’d prefer to use that more complete `FontForge.app` instead, but I haven’t had a chance to work through it yet. I don’t personally think it should block this update, since the existing port is broken, and this replaces it with a current and working version.

markmentovai replied:
I took a closer look at upstream’s `macbundle` target and it’s not so hot either.

So, in that it’s still better than nothing and not really any worse than the alternative, I borrowed your approach from #6289 to build a small shim app in the +gui variant.

This should be ready to merge now. Thanks!

markmentovai commented:
Pull request updated. Thanks!

mojca commented:
Looks good to me, thank you very much.
The explicit options are there to prevent opportunistic usage of dependencies (sometimes software would check whether a library exists and decide whether to use it based on that). It might not be absolutely necessary in this case (one would need to check what happens), but we prefer to be on the safe side just in case.

One final thing.
Fontforge appears to be a dependency of `lilypond` and `pdf2htmlex`. The second one is probably not widely used and may require a major upgrade (I seem to remember I wasn't able to upgrade it because it needed a newer fontforge), but at least for `lilypond` it might be worth checking if it requires some revbump and whether it still builds fine.
@nerdling @lemzwerg

mojca commented:
@udbraumann, @kencu: is anyone willing to check the build on some ancient hardware?
https://trac.macports.org/ticket/60409

kencu commented:
I am sure it will be broken -- I'll take on the fixup project afterwards.

I'm not sure about using their bundling plan in the cmake file -- it shoves the entire build into the app bundle, and we don't generally do that except in special circumstances. So I would not enable that plan. We'll come up with something better, like we did before.

lemzwerg commented:
> Fontforge appears to be a dependency of `lilypond`  [...]

Yes, it is.  And the next `lilypond-devel` version will need Python scripting.  Since python is already a requisite of lilypond I ask you to also update `lilypond-devel` so that it uses the `+python38` variant of fontforge.

markmentovai commented:
> Yes, it is. And the next `lilypond-devel` version will need Python scripting. Since python is already a requisite of lilypond I ask you to also update `lilypond-devel` so that it uses the `+python38` variant of fontforge.

It was a bit of a slog, but I included an update to the current `lilypond-devel` that I think does what you’re asking.

Like so?

(I’ve never used lilypond, so I wouldn’t really know if anything was out of place.)

lemzwerg commented:
> It was a bit of a slog, but I included an update to the current `lilypond-devel` that I think does what you’re asking.

This looks excellent, thanks!  BTW, this would be a good opportunity to get rid of the dblatex package dependency, too.  Could you use `reinplace` to do the equivalent of lilypond commit 250d841a?

markmentovai commented:
> > It was a bit of a slog, but I included an update to the current `lilypond-devel` that I think does what you’re asking.
> 
> This looks excellent, thanks! BTW, this would be a good opportunity to get rid of the dblatex package dependency, too. Could you use `reinplace` to do the equivalent of lilypond commit 250d841a?

I did this, and moved a few of the build-time dependencies to the `+docs` variant. (And added `ImageMagick`, which it has required since https://github.com/lilypond/lilypond/commit/cfa910f1c13ae760756692175a733db4e5a8899a, 8 years ago).

It all seems a little futile, though, since https://github.com/lilypond/lilypond/commit/19e4a54f40388e515ad099c9eda1c5909fd2d8df a year and a half ago pinned its `texi2html` requirement at 1.8.2 exactly. MacPorts hasn’t offered that version since f347735927fe updated it to 5.0, 10 years ago.

It seems that `lilypond-devel+docs` has been broken to varying degrees for a long time. I cleaned it up as requested (and then some), but it’s still broken, and I don’t really know if it’s worth fixing.

(I tested `lilypond-devel+docs` with this

```
    reinplace -W ${worksrcpath} \
        "s|\\(TEXI2HTML.*, 1\\.82\\), 1\\.82|\\1|g" \
        configure.ac
```

but don’t want to commit that because https://github.com/lilypond/lilypond/commit/19e4a54f40388e515ad099c9eda1c5909fd2d8df is pretty clear, with justification, about wanting to stay on 1.82.)

lemzwerg commented:
> > This looks excellent, thanks! BTW, this would be a good opportunity to get rid of the dblatex package dependency, too. Could you use `reinplace` to do the equivalent of lilypond commit 250d841a?
> 
> I did this, and moved a few of the build-time dependencies to the `+docs` variant.

Great!

> (And added `ImageMagick`, which it has required since [lilypond/lilypond@cfa910f](https://github.com/lilypond/lilypond/commit/cfa910f1c13ae760756692175a733db4e5a8899a), 8 years ago).

Thanks, but this is probably the wrong action right now, see below.

> It all seems a little futile, though, since [lilypond/lilypond@19e4a54](https://github.com/lilypond/lilypond/commit/19e4a54f40388e515ad099c9eda1c5909fd2d8df) a year and a half ago pinned its `texi2html` requirement at 1.8.2 exactly. MacPorts hasn’t offered that version since [f347735](https://github.com/macports/macports-ports/commit/f347735927fea4826b5b52c5e7fc5fc121a4f568) updated it to 5.0, 10 years ago.

For building the complete documentation it would be indeed necessary to create a texi2html-182 package.  I plan to eventually contribute it.

> It seems that `lilypond-devel+docs` has been broken to varying degrees for a long time. I cleaned it up as requested (and then some), but it’s still broken, and I don’t really know if it’s worth fixing.
> 
> (I tested `lilypond-devel+docs` with this
> 
> ```
>     reinplace -W ${worksrcpath} \
>         "s|\\(TEXI2HTML.*, 1\\.82\\), 1\\.82|\\1|g" \
>         configure.ac
> ```

Well, the `+docs` variant currently doesn't build the documentation at all!  It only builds the `.info` files, and for this action you neither need dblatex nor texi2html – nor ImageMagick.

The whole documentation for LilyPond gets only build if you call `make doc`.

lemzwerg commented on textproc/lilypond-devel/Portfile:
```diff
@@ -199,6 +205,13 @@ post-destroot {
 }
 
 variant docs description {Build documentation files} {
+    depends_build-append    port:ImageMagick \
```
As explained in https://github.com/macports/macports-ports/pull/7040#issuecomment-626916421, this is not necessary currently.

markmentovai replied:
Removed.

markmentovai commented:
> > > This looks excellent, thanks! BTW, this would be a good opportunity to get rid of the dblatex package dependency, too. Could you use `reinplace` to do the equivalent of lilypond commit 250d841a?
> > 
> > 
> > I did this, and moved a few of the build-time dependencies to the `+docs` variant.
> 
> Great!
> 
> > (And added `ImageMagick`, which it has required since [lilypond/lilypond@cfa910f](https://github.com/lilypond/lilypond/commit/cfa910f1c13ae760756692175a733db4e5a8899a), 8 years ago).
> 
> Thanks, but this is probably the wrong action right now, see below.

OK, I removed that portion. Please let me know if you’d like any other changes.

As I said, I’m not a LilyPond user, so I’m trying to help you out by doing the right thing, but to some extent I’m feeling my way around in the dark here.

Does everything look all right now?

markmentovai commented:
Is everyone satisfied with this update now?

lemzwerg commented:
Yes, LGTM.  Thanks for your work!
